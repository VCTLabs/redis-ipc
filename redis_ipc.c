#define _GNU_SOURCE // for gettid()
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <sys/types.h>
#include <sys/syscall.h>
#include <sys/time.h>
#include <json/json.h>
#include <hiredis/hiredis.h>
#include "redis_ipc.h"

#define safe_free(ptr) { if (ptr) free(ptr); ptr = NULL; }

struct redis_ipc_per_thread
{
//@@@@ FIXME: this will turn into a linked list
//  list_t list_head
    pid_t tid;               // owning thread
    char *component;         // component name used in redis
    char *thread;            // thread name used in redis
    char *result_queue_path; // based on component and thread
    redisContext *redis_state; // state for connection to redis server
};

static struct redis_ipc_per_thread *redis_ipc_info = NULL;

// gettid() is missing a libc wrapper for some reason
// (manpage even mentions it)
static pid_t gettid()
{
  return syscall(SYS_gettid);
}

struct redis_ipc_per_thread * get_per_thread_info()
{
    struct redis_ipc_per_thread *next_info = redis_ipc_info;
//@@@@@@ FIXME: switch to using a list, entry for each thread 
    //while (iterate through thread info list)
    if (next_info->tid == gettid())
    {
        return next_info; 
    }

    return NULL;
}

// NOTE: keep the following type enum and names array in sync

enum redis_ipc_type
{
    RPC_TYPE_INVALID = 0,
    RPC_TYPE_SETTING,
    RPC_TYPE_STATUS,
    RPC_TYPE_COMMAND,
    RPC_TYPE_RESULT,
    RPC_TYPE_EVENT,
    RPC_TYPE_DEBUG
};

const char *redis_ipc_type_names[] = 
{
   "INVALID",
   "settings",
   "status",
   "commands",
   "results",
   "events",
   "debug"
};

// generate redis key names aka our IPC "paths" 
static int ipc_path(char *buf, size_t buf_len, enum redis_ipc_type type, 
                    const char *component, 
                    // only used for "results" type
                    const char *thread, 
                    // only used for "events" type
                    const char *subchannel)
{
    int ret = RIPC_FAIL, path_len = -1;

    switch (type)
    {
        case RPC_TYPE_SETTING:
        case RPC_TYPE_STATUS:
            path_len = snprintf(buf, buf_len, "%s.%s", 
                                redis_ipc_type_names[type],
                                component);
            break;

        case RPC_TYPE_COMMAND:
            path_len = snprintf(buf, buf_len, "queues.%s.%s", 
                                component,
                                redis_ipc_type_names[type]);
            break;

        case RPC_TYPE_RESULT:
            if (thread != NULL)
            {
                path_len = snprintf(buf, buf_len, "queues.%s.%s.%s", 
                                    component, thread, 
                                    redis_ipc_type_names[type]);
            }
            break;

        case RPC_TYPE_EVENT:
        case RPC_TYPE_DEBUG:
            if (subchannel != NULL)
            {
                path_len = snprintf(buf, buf_len, "channel.%s.%s.%s", 
                                    component, subchannel, 
                                    redis_ipc_type_names[type]);
            }
            else
            {
                path_len = snprintf(buf, buf_len, "channel.%s.%s", 
                                    component, redis_ipc_type_names[type]);
            }
            break;
    }

    // path_len == buf_len indicates truncation
    if (path_len > 0 && path_len < buf_len)
    {
        ret = RIPC_OK;
    }

    return ret;
}

void cleanup_per_thread_info(struct redis_ipc_per_thread *thread_info)
{
    safe_free(thread_info->component);
    safe_free(thread_info->thread);
    safe_free(thread_info->result_queue_path);
    if (thread_info->redis_state)
    {
        // closes connection and frees memory related to this connection
        redisFree(thread_info->redis_state);
    }
}

// initialize per-thread state 
int redis_ipc_init(const char *this_component, const char *this_thread)
{
    struct redis_ipc_per_thread *new_info = NULL;
    char result_queue_path[RIPC_MAX_IPC_PATH_LEN];
    int ret = RIPC_FAIL;

    new_info = calloc(1, sizeof(struct redis_ipc_per_thread));

    // component is name for all threads/processes in same subsystem,
    // thread is uniquely (non-random) assigned name for this thread,
    new_info->component = strdup(this_component);
    new_info->thread = strdup(this_thread);
    if (new_info->component == NULL || new_info->thread == NULL)
        goto redis_ipc_init_failed;

    // tid is OS thread ID number
    new_info->tid = gettid();

    // each thread gets its own redis connection
    new_info->redis_state = redisConnect(RIPC_SERVER_IP, RIPC_SERVER_PORT);
    if (new_info->redis_state == NULL || new_info->redis_state->err)
        goto redis_ipc_init_failed;

    // each thread gets its own results queue in redis
    memset(result_queue_path, 0, sizeof(result_queue_path));
    ret = ipc_path(result_queue_path, sizeof(result_queue_path),
                  RPC_TYPE_RESULT, this_component, this_thread, NULL);
    if (ret != RIPC_OK)
        goto redis_ipc_init_failed;
    new_info->result_queue_path = strdup(result_queue_path);
    if (new_info->result_queue_path == NULL)
        goto redis_ipc_init_failed;

    redis_ipc_info = new_info; //@@@@@@ FIXME: switch to using a list, append entry for each new thread

    return RIPC_OK;

redis_ipc_init_failed:
    cleanup_per_thread_info(new_info);
    safe_free(new_info);

    return RIPC_FAIL;
}


int format_debug_msg(char *msg, size_t max_msg_len, 
                     const char *format, va_list argp)
{
    const char *trunc_warning = "[TRUNC]";
    size_t warning_len = strlen(trunc_warning);
    size_t warning_offset = max_msg_len - warning_len - 1;
    int msg_len = -1;
    int ret = RIPC_FAIL;

    // give up if message buffer is ridiculously small
    if (max_msg_len < warning_len+1) return RIPC_FAIL;

    msg_len = vsnprintf(msg, max_msg_len, format, argp);

    // flag truncation of message
    if (msg_len == max_msg_len)
    {
        // write warning over end of message
        strncpy(&msg[warning_offset], trunc_warning, warning_len);
    }

    if (msg_len > 0)
    {
        ret = RIPC_OK;
    }

    return ret;
}

//@@@@ FIXME: debug will be dynamically configurable from a setting or config file
int get_debug_verbosity()
{
    return 5;
}

//@@@@ FIXME: debug will be dynamically configurable from a setting or config file
int stderr_debug_is_enabled()
{
    return 1;
}

// run redis command and reset connection upon error
//
// if return value is non-null,
// caller is responsible for calling freeReplyObject() when done with it
redisReply * redis_command(const char *format, ...)
{
    struct redis_ipc_per_thread *thread_info = get_per_thread_info();
    redisReply *reply = NULL;
    va_list argp;
    
    if (stderr_debug_is_enabled())
    {
        va_start(argp, format);
        vfprintf(stderr, format, argp);
        fprintf(stderr, "\n");
        va_end(argp);
    }

    va_start(argp, format);
    reply = redisvCommand(thread_info->redis_state, format, argp);
    va_end(argp);

    if (reply == NULL)
    {
        // must reconnect to redis server after an error 
        redisFree(thread_info->redis_state);
        thread_info->redis_state = redisConnect(RIPC_SERVER_IP, RIPC_SERVER_PORT);
    }

    return reply;
}

// standard fields for json objects that will be useful when
// troubleshooting component integration issues, if nothing else...
#define MAX_TIMESTAMP_LEN 64 
void json_add_common_fields(json_object *obj)
{
    struct redis_ipc_per_thread *thread_info = get_per_thread_info();
    char timestamp_buffer[MAX_TIMESTAMP_LEN];
    struct timeval timestamp;

    // collect timestamp as string
    memset(&timestamp, 0, sizeof(timestamp));
    memset(timestamp_buffer, 0, sizeof(timestamp_buffer));
    gettimeofday(&timestamp, NULL);
    snprintf(timestamp_buffer, sizeof(timestamp_buffer), "%ld.%06ld", 
             timestamp.tv_sec, timestamp.tv_usec);

    json_object_object_add(obj, "timestamp", 
                           json_object_new_string(timestamp_buffer));
    json_object_object_add(obj, "component", 
                           json_object_new_string(thread_info->component));
    json_object_object_add(obj, "thread", 
                           json_object_new_string(thread_info->thread));
    json_object_object_add(obj, "tid", json_object_new_int(gettid()));
}

static int redis_publish(const char *channel_path, json_object *obj)
{
    const char *json_text = NULL;
    int ret = RIPC_FAIL;
    redisReply *reply = NULL;

    // append channel path
    json_object_object_add(obj, "channel", 
                           json_object_new_string(channel_path));

    // append standard fields (such as timestamp)
    json_add_common_fields(obj);

    json_text = json_object_to_json_string(obj);

    // don't forget to free reply
    reply = redis_command("PUBLISH %s %s", channel_path, json_text);

    if (reply != NULL)
    {
        ret = RIPC_OK;
        freeReplyObject(reply);
    }

    return ret;
}

// caller is responsible for cleaning up the message object
int redis_ipc_send_event(const char *subchannel, json_object *message)
{
    char event_channel_path[RIPC_MAX_IPC_PATH_LEN];
    struct redis_ipc_per_thread *thread_info = get_per_thread_info();
    int ret = RIPC_FAIL;

    // calculate channel name
    ret = ipc_path(event_channel_path, sizeof(event_channel_path),
                   RPC_TYPE_EVENT, thread_info->component, NULL, subchannel);
    if (ret != RIPC_OK)
        goto redis_ipc_send_event_finish;

    // publish event object
    ret = redis_publish(event_channel_path, message);

redis_ipc_send_event_finish:

   return ret;
}


int redis_ipc_send_debug(unsigned int debug_level, const char *format, ...)
{
    char msg_buffer[RIPC_MAX_DEBUG_LEN];
    char debug_channel_path[RIPC_MAX_IPC_PATH_LEN];
    va_list argp;
    struct redis_ipc_per_thread *thread_info = get_per_thread_info();
    json_object *debug_obj = NULL;
    int ret = RIPC_FAIL;

    // ignore if the debug level is higher than current verbosity
    if (debug_level > get_debug_verbosity())
    {
        return RIPC_OK;  // it's easy to successfully do nothing...
    }

    // calculate channel name
    ret = ipc_path(debug_channel_path, sizeof(debug_channel_path),
                   RPC_TYPE_DEBUG, thread_info->component, NULL, NULL);
    if (ret != RIPC_OK)
        goto redis_ipc_send_debug_finish;

    // format message into string
    memset(msg_buffer, 0, sizeof(msg_buffer));
    va_start(argp, format);
    ret = format_debug_msg(msg_buffer, sizeof(msg_buffer), format, argp);
    va_end(argp);
    if (ret != RIPC_OK)
        goto redis_ipc_send_debug_finish;

    // generate debug object
    debug_obj = json_object_new_object();
    if (debug_obj == NULL)
        goto redis_ipc_send_debug_finish;
    json_object_object_add(debug_obj, "message", 
                           json_object_new_string(msg_buffer));
    json_object_object_add(debug_obj, "level", 
                           json_object_new_int(debug_level));

    // publish debug object
    ret = redis_publish(debug_channel_path, debug_obj);

redis_ipc_send_debug_finish:
   json_object_put(debug_obj);  // safe to call even if NULL

   return ret;
}

static int redis_subscribe(const char *channel_path)
{
    int ret = RIPC_FAIL;
    redisReply *reply = NULL;

    // don't forget to free reply
    reply = redis_command("PSUBSCRIBE %s", channel_path);

    if (reply != NULL)
    {
        ret = RIPC_OK;
        freeReplyObject(reply);
    }

    return ret;
}

// if component is NULL, subscribe to all event channels
// if only subchannel is NULL, subscribe to all event channels for component
// (for some components there may only be one event channel)
int redis_ipc_subscribe_events(const char *component, const char *subchannel)
{
    char event_channel_path[RIPC_MAX_IPC_PATH_LEN];
    char *component_pattern = NULL;
    size_t pattern_len = 0;
    int ret = RIPC_FAIL;

    // calculate channel name
    if (component == NULL) 
    {
        // use wildcard to get all event channels
        component_pattern = strdup("*");
    }
    else
    {
        if (subchannel == NULL)
        {
            // append wildcard to get all event channels for this component
            pattern_len = strlen(component)+2;
            component_pattern = calloc(1, pattern_len);
            snprintf(component_pattern, pattern_len, "%s%c", component, '*');
        }
        else
        {
            // exact match for both component and subchannel
            component_pattern = strdup(component);
        }
    }
    ret = ipc_path(event_channel_path, sizeof(event_channel_path),
                   RPC_TYPE_EVENT, component_pattern, NULL, subchannel);
    if (ret != RIPC_OK)
        goto redis_ipc_subscribe_events_finish;

    // publish event object
    ret = redis_subscribe(event_channel_path);

redis_ipc_subscribe_events_finish:
   safe_free(component_pattern);

   return ret;
}

int redis_ipc_subscribe_debug(const char *component)
{
    char debug_channel_path[RIPC_MAX_IPC_PATH_LEN];
    const char *component_pattern = NULL;
    size_t pattern_len = 0;
    int ret = RIPC_FAIL;

    // calculate channel name
    if (component == NULL) 
    {
        // use wildcard to get all debug channels
        component_pattern = "*";
    }
    else
    {
        // exact match 
        component_pattern = component;
    }
    ret = ipc_path(debug_channel_path, sizeof(debug_channel_path),
                   RPC_TYPE_DEBUG, component_pattern, NULL, NULL);
    if (ret != RIPC_OK)
        goto redis_ipc_subscribe_debug_finish;

    // publish debug object
    ret = redis_subscribe(debug_channel_path);

redis_ipc_subscribe_debug_finish:

   return ret;
}

static int redis_unsubscribe(char *channel_path)
{
    redisReply *reply = NULL;
    int ret = RIPC_FAIL;

    // don't forget to free reply
    reply = redis_command("PUNSUBSCRIBE %s", channel_path);

    if (reply != NULL)
    {
        ret = RIPC_OK;
        freeReplyObject(reply);
    }

    return ret;
}

// unsubscribe from all event channels
int redis_ipc_unsubscribe_events()
{
    char event_channel_path[RIPC_MAX_IPC_PATH_LEN];
    int ret = RIPC_FAIL;

    ret = ipc_path(event_channel_path, sizeof(event_channel_path),
                   RPC_TYPE_EVENT, "*", NULL, NULL);
    if (ret != RIPC_OK)
        goto redis_ipc_unsubscribe_events_finish;

    ret = redis_unsubscribe(event_channel_path);

redis_ipc_unsubscribe_events_finish:

   return ret;
}

// unsubscribe from all debug channels
int redis_ipc_unsubscribe_debug()
{
    char debug_channel_path[RIPC_MAX_IPC_PATH_LEN];
    int ret = RIPC_FAIL;

    ret = ipc_path(debug_channel_path, sizeof(debug_channel_path),
                   RPC_TYPE_DEBUG, "*", NULL, NULL);
    if (ret != RIPC_OK)
        goto redis_ipc_unsubscribe_debug_finish;

    ret = redis_unsubscribe(debug_channel_path);

redis_ipc_unsubscribe_debug_finish:

   return ret;
}

// current thread should be subscribed to one or more callers before calling
// caller is responsible for cleaning up the returned message object
json_object * redis_ipc_get_message_blocking(void)
{
    json_object *message = NULL;
    redisReply *reply = NULL;
    const char *message_str = NULL;
    struct redis_ipc_per_thread *thread_info = get_per_thread_info();
    int ret = RIPC_FAIL;

    // block until a message is available
    ret = redisGetReply(thread_info->redis_state, &reply);
    if (ret != REDIS_OK)
    {
        // must reconnect to redis server after an error 
        redisFree(thread_info->redis_state);
        thread_info->redis_state = redisConnect(RIPC_SERVER_IP, RIPC_SERVER_PORT);

        goto redis_get_channel_message_finish;
    }


    // extract message from reply object
    //
    // reply should be an array: 
    //   string "pmessage" 
    //   string <psubscribe pattern>
    //   string <sending channel>
    //   string <message> ** entry of interest
    if (reply->type != REDIS_REPLY_ARRAY)
        goto redis_get_channel_message_finish;
    if (reply->element[3]->type != REDIS_REPLY_STRING)
        goto redis_get_channel_message_finish;

    message_str = reply->element[3]->str;

    if (stderr_debug_is_enabled())
    {
        fprintf(stderr, "MESSAGE %s\n", message_str);
    }

    // parse message back into json object
    message = json_tokener_parse(message_str);
    if (is_error(message)) message = NULL;

redis_get_channel_message_finish:
    if (reply != NULL)
        freeReplyObject(reply);

   return message;
}
