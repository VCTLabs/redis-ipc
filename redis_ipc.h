
#ifndef __REDIS_IPC_H__
#define __REDIS_IPC_H__

#include <sys/types.h>
#include <json/json.h>

//*************
// constants
//*************

// return codes > 0 are also OK
#define RIPC_OK     0
#define RIPC_FAIL (-1)

// location of config file
//#define RIPC_CONF_PATH "/etc/redis_ipc.conf"
#define RIPC_CONF_PATH "./redis_ipc.conf"

//*********** things that might move to config file when it is implemented

// max length of formatted debug message (will be truncated if longer)
#define RIPC_MAX_DEBUG_LEN 1024

// max length of constructed redis key name
#define RIPC_MAX_IPC_PATH_LEN 128

// redis server address
#define RIPC_SERVER_IP "127.0.0.1"
#define RIPC_SERVER_PORT 6379

//*************
// functions
//*************

// Init should be called from each thread that will be doing redis IPC,
// with all threads getting same component name but unique thread name.
// Cleanup is provided to prevent a slow leak from turnover of short-lived
// threads that each call init to allocate a few per-thread variables.
// Cleanup is done based on thread ID so that main thread can clean up for
// terminated threads.
//
// The thread name should indicate the purpose of the thread and be 
// predictable, e.g. live-http-worker-3, rather than randomly generated or
// based on TID number, because it will be used to generate name of results
// queue -- and we want same results queues to be re-used if component gets
// restarted (avoid having to garbage-collect stale, abandoned redis queues)
int redis_ipc_init(const char *this_component, const char *this_thread);
int redis_ipc_cleanup(pid_t tid);

// A component can send a command to any other component, 
// but it can only receive commands from its own command queue(s).
//
// The timeout field is in seconds, or use zero to indicate no timeout
// for blocking.
//
// When submitting a command, the name of submitter's results queue 
// and the command ID will automatically be inserted into the command. 
//
// The submitter will block on its results until reply has been received 
// or timeout has been exceeded. If a reply with _different_ command ID
// is received, it will be logged as an error to submitter component's 
// debug channel then freed, and the wait on results queue will restart
// (expiration time will be reset to original).
//
// After receiving and executing a command, the receiving component
// should submit a reply. The original command is passed in as a parameter
// to provide the command ID and path to results queue. The command ID
// will automatically be added to the reply object before pushing it to the
// result queue so that the submitter can verify that this reply is the
// expected one (as opposed to a stale reply, belonging to a command that 
// took so long that submitter timed out before seeing the reply).
json_object * redis_ipc_send_command_blocking(const char *dest_component, 
                                              const char *queue_name, 
                                              unsigned int command_id, 
                                              json_object *command, 
                                              unsigned int timeout);
json_object * redis_ipc_receive_command_blocking(const char *queue_name,
                                              unsigned int timeout);
int redis_ipc_send_reply(const json_object *completed_command, json_object *reply);
                        

// A component can only write a setting if it has been authorized by redis_ipc.conf, 
// but it can read any setting.
//
// If multiple status fields are being accessed, probably should access the full thing
// rather than multiple calls to individual fields.
int redis_ipc_write_setting(const char *owner_component, json_object *value);
json_object * redis_ipc_read_setting(const char *owner_component);
int redis_ipc_write_setting_field(const char *owner_component, const char *field_name, json_object *value);
json_object * redis_ipc_read_setting_field(const char *owner_component, const char *field_name);

// A component can only write its own status, 
// but it can read any status.
//
// If multiple status fields are being accessed, probably should access the full thing
// rather than multiple calls to individual fields.
int redis_ipc_write_status(json_object *value);
json_object * redis_ipc_read_status(const char *owner_component);
int redis_ipc_write_status_field(const char *field_name, json_object *value);
json_object * redis_ipc_read_status_field(const char *owner_component, const char *field_name);

// Each component will send debug messages to its own debug channel,
// but can subscribe to any (or all) debug channels.
//
// A component sending an event should supply NULL as subchannel unless
// it has multiple subchannels defined in redis_ipc.conf,
// in which case it should supply one of those subchannels. The name of
// the channel to which an event is posted will automatically get added 
// to the message as "source" field so that subscribers to multiple 
// channels can easily determine which channel a message came from.
//
// A component subscribing to events can use NULL as component parameter
// to watch all event channels for all components. Or, to watch all events
// from a single component, use NULL for subchannel (as mentioned above,
// not all components even have subchannels).
// 
// A thread watching for events will block indefinitely until the next
// event is received, because redis protocol does not implement a timeout
// on waiting for messages. 
//
// The unsubscribe function stops watching all event channels.
int redis_ipc_send_event(const char *subchannel, json_object *message);
int redis_ipc_subscribe_events(const char *component, const char *subchannel);
json_object * redis_ipc_watch_events_blocking();
int redis_ipc_unsubscribe_events();

// Each component will send debug messages to its own debug channel,
// but can subscribe to any (or all) debug channels. Use NULL as
// component to subscribe to all debug channels.
//
// Debug send can serve as drop-in replacement for printf-style logging,
// which is why it doesn't take a JSON object. The send function internally
// generates a json object containing a "message" field and a "timestamp" 
// field. The component and thread posting the message will automatically be 
// added to the object so that subscribers to multiple debug channels can 
// easily determine who sent the message.

// When sending debug messages, low debug level indicates
// high message priority, since only messages of lower or equal level
// to configured component debug verbosity will actually get sent.
// 
// A thread watching for debug messages will block indefinitely until the next
// debug message is received, because redis protocol does not implement 
// a timeout on waiting for messages. 
//
// The unsubscribe function stops watching all debug channels.
int redis_ipc_send_debug(unsigned int debug_level, const char *format, ...);
int redis_ipc_subscribe_debug(const char *component);
json_object * redis_ipc_receive_debug_blocking();
int redis_ipc_unsubscribe_debug();
#endif
