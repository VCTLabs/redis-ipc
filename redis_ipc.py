# This is a Python module to provide client access to the Redis server
# it is treated as a library

import redis
import time
import pdb   ########## debug  ( no kidding)
# global data
"""
"""

# global functions
def is_json_msg(m):   # do we need this???
    global component
    pass
 
# exceptions
class RedisIpcExc(Exception):
    pass
NoRedis     =  RedisIpcExc("redis server not available")
BadMessage  =  RedisIpcExc("redis message not a recognizable message")
MsgTimeout  =  RedisIpcExc("redis message request timed out")

# the main feature here is a class which will provide the wanted access
class redis_client(object):
    
    def __init__(self,component):
        # component : queue name for listening
        self.component=component
        # @@@@@@@@@@@@(process number of this component)
        self.process_number=str(1234)

    def generate_msg_id(self):
        # unique id for message
        # component name, process number, timestamp
        ts=str(time.time())   # floating timestamp
        proc=99   # @@@@@@@@@@@@@@@@
        the_id=self.component + ':' + self.process_number + ':' + ts
        return the_id
        
    def redis_ipc_send_and_receive(self,dest,cmd,tmout):
        """
        dest     -   the name of a component queue (string)
        cmd      -   the command to send (a Python dictionary)
        tmout    -   timeout for receiving a response, floating seconds
                     this method calls the method 
                     redis_ipc_receive_command_blocking
                     with this timeout
                     
        calls the (unblocking) redis_ipc_send_command
        for the actual send
        """
        self.redis_ipc_send(dest, cmd)
        
    def redis_ipc_send(self,dest, cmd):
        """
        arguments are mandatory
        dest     -   queue name for recipient queue
        cmd      -   a command known to the receiving component
                     it is a JSON dictionary
                     
        this routine does not block
        it just sends the command to the back of the queue
        """
        # the command is a Python dictionary
        # turn it into a JSON dictionary before sending it
        # send it via Redis
        pass

    def redis_ipc_receive_command_blocking(self,qq,tmout):
        """
        arguments are mandatory
        qq the components receive queue
        tmout   -   timeout for receiving a response, floating seconds
        
        the response is a JSON dictionary
        turn it into a Python dictionary
        return it
        """
        pass
        
   
        
    