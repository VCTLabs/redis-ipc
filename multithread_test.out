# sample output from multithread_test
PUBLISH channel.debug.printer { "message": "printer starting to smoke", "level": 1, "channel": "channel.debug.printer", "timestamp": "1401058399.161286", "component": "printer", "thread": "monitor", "tid": 20152 }
PUBLISH channel.debug.button { "message": "NO don't press that button...", "level": 1, "channel": "channel.debug.button", "timestamp": "1401058399.161285", "component": "button", "thread": "watcher", "tid": 20153 }
PUBLISH channel.debug.printer { "message": "printer on fire!!", "level": 0, "channel": "channel.debug.printer", "timestamp": "1401058399.161965", "component": "printer", "thread": "monitor", "tid": 20152 }
PUBLISH channel.debug.button { "message": "I told you not to press it!!", "level": 0, "channel": "channel.debug.button", "timestamp": "1401058399.162194", "component": "button", "thread": "watcher", "tid": 20153 }
