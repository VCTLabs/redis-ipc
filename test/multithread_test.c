#define _GNU_SOURCE // for gettid()
#include <sys/types.h>
#include <unistd.h>
#include <sys/syscall.h>
#include <pthread.h>
#include "redis_ipc.h"

// gettid() is missing a libc wrapper before glibc 2.30
// (manpage even mentions it)
#if (__GLIBC_MINOR__ < 30)
pid_t gettid()
{
  return syscall(SYS_gettid);
}
#endif

void *run_printer_thread(void *data)
{
  redis_ipc_init("printer", "monitor");
  redis_ipc_send_debug(1, "printer starting to smoke");
  redis_ipc_send_debug(0, "printer on fire!!");
  redis_ipc_cleanup(gettid());

  return 0;
}

void *run_button_thread(void *data)
{
  redis_ipc_init("button", "watcher");
  redis_ipc_send_debug(1, "NO don't press that button...");
  redis_ipc_send_debug(0, "I told you not to press it!!");
  redis_ipc_cleanup(gettid());

  return 0;
}

int main(int argc, char **argv)
{
  pthread_t printer_thread_info, button_thread_info;

  pthread_create(&printer_thread_info, NULL, run_printer_thread, NULL);
  pthread_create(&button_thread_info, NULL, run_button_thread, NULL);

  pthread_join(printer_thread_info, NULL);
  pthread_join(button_thread_info, NULL);

  return 0;
}
