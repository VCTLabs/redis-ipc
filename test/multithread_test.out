(button) CONFIG SET notify-keyspace-events Kh
(printer) CONFIG SET notify-keyspace-events Kh
(button) PUBLISH channel.debug.button { "message": "NO don't press that button...", "level": 1, "channel": "channel.debug.button", "timestamp": "1668592241.554110", "component": "button", "thread": "watcher", "tid": 19190 }
(printer) PUBLISH channel.debug.printer { "message": "printer starting to smoke", "level": 1, "channel": "channel.debug.printer", "timestamp": "1668592241.554110", "component": "printer", "thread": "monitor", "tid": 19189 }
(printer) PUBLISH channel.debug.printer { "message": "printer on fire!!", "level": 0, "channel": "channel.debug.printer", "timestamp": "1668592241.554176", "component": "printer", "thread": "monitor", "tid": 19189 }
(button) PUBLISH channel.debug.button { "message": "I told you not to press it!!", "level": 0, "channel": "channel.debug.button", "timestamp": "1668592241.554177", "component": "button", "thread": "watcher", "tid": 19190 }
